### ISSUE TITLE
training step

### ISSUE THREAD

**Issue by Michaela1224** [cornell-zhang/FracBNN#3]
Hi,
 I'm trying to reproduce your work. But I have some problems. You use a two-step training in the README, but Step 3 is mentioned in your paper: "To train the FracBNN, the first two steps are the same except that the activations are quantized to 2 bits. We add a third step that initializes the weights from the second step, and applies the fractional convolutional layer". According to the paper, what is the difference between the "activation quantization to 2 bits" in the first two steps and the PGBinaryConv2d function?  According to README, is the function "PGBinaryConv2d" used in each step of the two-step training? The training tricks of BNN are important, I am not sure whether I understand correctly, looking forward to your reply.

**Comment by ychzhang** [comment]
Hi @yanggeng1234 ,

Thanks for checking out our work. The 3-step training documented in the paper is what we did in our experiments. PGBinaryConv2d is only used in the 3rd step of the training.
For the first two steps, we use the regular "BinaryConv2d" module for the convolutional layers. Note that we quantize the inputs to 2 bits using the q.QuantSign module. That's what we meant by "the first two steps are the same except that the activations are quantized to 2 bits".
For the third step, we replace the "BinaryConv2d" module with the "PGBinaryConv2d", but inheriting the same weights. Then train the model again.
Note that the PG layer is simulating the fractional bitwidth on hardware. Though it takes 2-bit activations as inputs, it does not utilize all input bits.
Hope this would help.

Yichi

**Comment by yanggeng1234** [comment]
Hi Yichi Zhang, 
Thanks for your reply. I retrained FracBNN used Cifar10 as you described in detail. Finally, the accuracy of 88.5% (step 1: 88.5% -&gt; step 2: 88.6%-&gt; step 3: 88.5%) is obtained, which is different from 89.1 described in your paper. And I found that the new results and the accuracy obtained by the two-step training I used previously (PGBinaryConv2d function is used for both steps) were 0.3 % higher. I expect to obtain the result closer to that of the paper. I attached my detailed settings. I just started contacting the BNN design. Looking forward to your better advice, and I'm continuing to experiment with other training Settings.


GengYang


------------------ 原始邮件 ------------------
发件人:                                                                                                                        "cornell-zhang/FracBNN"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年9月29日(星期三) 凌晨2:03
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [cornell-zhang/FracBNN] training step (#3)





 
Hi @yanggeng1234 ,
 
Thanks for checking out our work. The 3-step training documented in the paper is what we did in our experiments. PGBinaryConv2d is only used in the 3rd step of the training.
 For the first two steps, we use the regular "BinaryConv2d" module for the convolutional layers. Note that we quantize the inputs to 2 bits using the q.QuantSign module. That's what we meant by "the first two steps are the same except that the activations are quantized to 2 bits".
 For the third step, we replace the "BinaryConv2d" module with the "PGBinaryConv2d", but inheriting the same weights. Then train the model again.
 Note that the PG layer is simulating the fractional bitwidth on hardware. Though it takes 2-bit activations as inputs, it does not utilize all input bits.
 Hope this would help.
 
Yichi
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by Michaela1224** [comment]
Thanks for your reply. I retrained FracBNN used Cifar10 as you described in detail. Finally, the accuracy of 88.5% (step 1: 88.5% -> step 2: 88.6%-> step 3: 88.5%) is obtained, which is different from 89.1 described in your paper. And I found that the new results were 0.3% higher than the accuracy obtained by the two-step training I used previously (PGBinaryConv2d function is used for both steps). I expect to obtain a result closer to that of the paper. Looking forward to your better advice, and I'm continuing to experiment with other training Settings for better results. 
My training setting is as follows:
step 1 : Run python cifar10.py -e 300 -wd 1e-5 -gpu 0 -s  (I used "BinaryConv2d" for all convolutional layers, and deleted the related loss about “gating target threshold" in Cifar10.py)
step 2:  Run python cifar10 -e 300 -wd 0 -f -r save_CIFAR10_model/model_binput-pg.pt -s -gpu 0
step 3: Run python cifar10 -g 0.0 -e 300 -wd 0 -f -r save_CIFAR10_model/model_binput-pg-finetune.pt -s -gpu 0

**Comment by yanggeng1234** [comment]
> Hi @yanggeng1234 ,
> 
> Thanks for checking out our work. The 3-step training documented in the paper is what we did in our experiments. PGBinaryConv2d is only used in the 3rd step of the training. For the first two steps, we use the regular "BinaryConv2d" module for the convolutional layers. Note that we quantize the inputs to 2 bits using the q.QuantSign module. That's what we meant by "the first two steps are the same except that the activations are quantized to 2 bits". For the third step, we replace the "BinaryConv2d" module with the "PGBinaryConv2d", but inheriting the same weights. Then train the model again. Note that the PG layer is simulating the fractional bitwidth on hardware. Though it takes 2-bit activations as inputs, it does not utilize all input bits. Hope this would help.
> 
> Yichi

**Comment by ychzhang** [comment]
It seems like 88.8% is currently the best you can get, if I understand it correctly. Your training steps look correct to me.
What I can suggest is: try to add one more training step with a lower initial learning rate, e.g., 1e-5, or a different gating target in the 3rd step, e.g., 3.

**Comment by Michaela1224** [comment]
Thank you again for your detailed reply！
